### PR TITLE
DM-32097: initial prototype of image cutout service backend.

### PR DESCRIPTION
**Pull request by TallJimbo**
There are some tests I'd still like to write, but I think this is ready for review.

The high-level entry point is the `ImageCutoutBackend` class - the idea is that you'll have an instance of one these started when the service starts, make one of the stencil classes from the `stencils` module with the geometry you want, and then call `ImageCutoutBackend.process_uuid`, at least for the UUID-oriented spec from the Jira ticket.  There are other `process_*` methods for other viable inputs that we might want to consider in the future.

### CODE REVIEW

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (rra) [inline python/lsst/image_cutout_backend/_image_cutout_backend.py:214]: I don't understand this parameter. Doesn't the UUID uniquely specify the dataset to use as the source?

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (rra) [inline python/lsst/image_cutout_backend/_image_cutout_backend.py]: I assume this is supposed to be the Butler from which the original image is read, not from which the *cutouts* are read.

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (rra) [inline python/lsst/image_cutout_backend/_image_cutout_backend.py:132]: Why is this required? I'd rather not write the files to local disk because then I have to worry about available space on the local disk.

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (rra) [inline python/lsst/image_cutout_backend/_image_cutout_backend.py]: If we're not using the Butler to store files, please don't do any of this; it just makes URLs more complicated to no real purpose that I can see. Instead, just write the output into a unique file name (probably via a UUID) and return where it was written. I don't see anything to be gained from encoding the dataset identifier and cutout parameters into the file name.

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (rra) [inline python/lsst/image_cutout_backend/_image_cutout_backend.py]: I don't like the use of "read" here because to me it implies that the cutout has already been calculated and stored and we're just reading the existing file from storage, and my understanding is that this is creating the cutout. Either that, or if this is reading an existing cutout from storage, I have entirely misunderstood how all of this is supposed to work and may need more orientation.

File: python/lsst/image_cutout_backend/stencils.py
Comment (rra) [inline python/lsst/image_cutout_backend/stencils.py:316]: I assume this will be fixed in a future version?

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (rra) [inline python/lsst/image_cutout_backend/_image_cutout_backend.py:125]: How is the frontend supposed to create this parameter?

**Review by rra — COMMENTED** [review]
The intent of the division of responsibility here is that the backend is a black box from the frontend's perspective, so in that sense I'm okay with whatever implementation strategy you choose, but there are a couple of design constraints here that I want to make sure you've thought of.

First, cutouts will be supported on virtual data products, so the frontend may be calling this code with a UUID that identifies a virtual data product that has to be recreated on the fly in order to take a cutout from it. That was one of the reasons why we were thinking about building this on top of pipelines. Maybe there's some better way to handle that, but just be aware that the frontend is not going to recreate the virtual data product in advance. This code will just get an identifier for a virtual data product and will have to cope.

Second, we will eventually need to get results in other image formats, so this backend will need to support converting to JPEG or whatever people end up wanting. It wasn't clear to me how that was plumbed through in this design.

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (timj) [reply python/lsst/image_cutout_backend/_image_cutout_backend.py:214]: The UUID uniquely specifies the dataset from the Registry point of view. The datastore can split a single python type into multiple files (for example storing the image plane and variance plane as separate files). To uniquely access a specific file in datastore you might need the UUID and component.

Comment (rra) [reply python/lsst/image_cutout_backend/_image_cutout_backend.py:214]: Ah, thanks. I think my question is then how do we anticipate this information being plumbed through DataLink and ObsTAP? The design here is that the cutout frontend doesn't know anything about the structure of identifiers; they're just opaque identifiers that it gets from an SIA or ObsTAP query and that it passes to the backend. So if we need something besides the UUID, we're back to needing a serialization format that SIA and ObsTAP will use so that the cutout service gets a serialized identifier that can be passed to the backend.

Comment (timj) [reply python/lsst/image_cutout_backend/_image_cutout_backend.py:214]: ObsTAP and data link don't know about composite datasets so SODA isn't going to know about it either unless some how there's a way for a SODA user to ask for a cutout of the variance plane without wanting the image plane.

Comment (rra) [reply python/lsst/image_cutout_backend/_image_cutout_backend.py:214]: Ah, okay, so basically I should ignore this parameter because the cutout service probably won't use it (at least unless we develop some local extension out of need later on)?

Comment (TallJimbo) [reply python/lsst/image_cutout_backend/_image_cutout_backend.py:214]: Yeah, I think you can ignore this for now.  It could be a useful extension for users who want to get a lot of cutouts and are concerned about the total overall size, because the common case (just getting the main image plane) can save a factor of 2 and change, but it's not any kind of immediate concern.

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (TallJimbo) [reply python/lsst/image_cutout_backend/_image_cutout_backend.py:132]: It's because all of our low-level FITS I/O code goes through the cfitsio C library, and that usually operates on regular C file pointers, so we don't have a direct object store write (or read) for any of these types; this local temporary file stuff is what happens behind Butler for these objects, too.  It should be possible to tell cfitsio to write to a contiguous block of memory that we've allocated for it instead, and then send those bytes directly to an object store, but we'll need to do some plumbing in obs_base.  For large cutouts, it also means you've traded a free filesystem space question, if that matters (though maybe the filesystem one is actually worse because the temporaries aren't freed up as reliably?).

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (TallJimbo) [reply python/lsst/image_cutout_backend/_image_cutout_backend.py]: How about "extract" or "extract_from" instead of "read"?  Your understanding of what it is doing is correct; I said "read" because I'm used to thinking of the read operation as what actually creates the cutout, because I'm used to thinking of cutouts as in-memory things, not persisted things; between this and the comment on the butler `__init__` parameter docs, I think the lesson here is that I need to be more consistent about "cutout" meaning or implying the persisted thing only, since that's the real deliverable.

File: python/lsst/image_cutout_backend/stencils.py
Comment (TallJimbo) [reply python/lsst/image_cutout_backend/stencils.py:316]: Yes.  It'll require some moderately tricky geometry code that I'm not looking forward to writing, but it's clear we need to do it; I just drew the boundary for this ticket at things I could do with our existing low-level geometry code.

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (TallJimbo) [reply python/lsst/image_cutout_backend/_image_cutout_backend.py:125]: For the foreseeable future, `ProjectionFinder.make_default()` will return something that works, and I'm comfortable hard-coding that initially - it's entirely conceivable it will never have to change.

This essentially encodes conventions about what dataset types our pipelines produce and how they are related, and I wasn't comfortable putting that kind of information directly into the `ImageCutoutBackend` class.  But I also wasn't sure if the long term fix is to put these conventions in a configuration file that would be packaged with the service, or to find a way to make this introspectable from the data repository itself.

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (TallJimbo) [reply python/lsst/image_cutout_backend/_image_cutout_backend.py]: Correct, and I think that's a sign I should adjust the docs to avoid "cutout" here.

**Comment by TallJimbo** [comment]
I did think some about virtual data products and non-FITS cutouts (more so the former) and I think there are a lot of open questions that are very much worth discussing (again, mostly about the former).  But beyond that I think I'll take @frossie up on her suggestion of a live call, as I think that's a better idea than me spending time writing prose on a DMTN-208 PR, as I previously suggested, or writing a novel in PR comments here.  I'm just going to write some quick slides with bullets to help organize my thoughts tomorrow, and then hopefully we can figure out the right participants and schedule something.

Overall, some of the decisions I've made here that may have been surprising are things I'm pretty confident in (e.g. not making the cutout backend a `PipelineTask`), but I had to fill in a lot of blanks about virtual datasets with assumptions to get there, and I know I have some philosophical biases about butler concepts that should probably be subjected to scrutiny.

Other choices here are not really decisions so much as placeholders, where I know what I've written will need to be replaced, but I don't have a solid answer for what to replace it with (not using `Butler` or even `Datastore` for writes).

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (rra) [reply python/lsst/image_cutout_backend/_image_cutout_backend.py:132]: The problem isn't so much cleaning up temporary space as that it's often easier to allocate memory than disk in cloud environments. But it sounds like it's inherent in the libraries we're using. I don't think it's worth going to a lot of effort to change until we see whether it's a problem.

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (rra) [reply python/lsst/image_cutout_backend/_image_cutout_backend.py]: extract sounds great. That makes sense to me given my mental model of what the Butler is doing under the hood. Thanks!

File: python/lsst/image_cutout_backend/_image_cutout_backend.py
Comment (rra) [reply python/lsst/image_cutout_backend/_image_cutout_backend.py:125]: That sounds good. Thanks!

**Comment by rra** [comment]
I think it's fine to not use the Butler to store the output, the more that I think about it. The implication is that the cutout service will get its own GCS bucket that it will use to store results, manage its own signed links to that bucket, and then just set a bucket expiration policy so that cutouts will be automatically deleted after some period of time (currently seven days).

The main implication of this is that the read policy will then just be that if you created the cutout, you get to read it, otherwise you don't, so anyone who wants to make the cutout available to anyone else will have to download it and store it somewhere more accessible. Sharing cutout URLs will not be possible because they'll be authenticated and tied to a specific user (not a group).

If that's not okay, then we should revisit using the Butler; I'm pretty strongly opposed to having more than one storage API that does full-blown group-based access control or attempting to replicate the authorization model in each VO service. That was intended to be the job of the client/server Butler.

**Comment by rra** [comment]
I've now rewritten the interface in vo-cutouts to use this new backend. Could you review the file `workers.py` at:

https://github.com/lsst-sqre/vo-cutouts/pull/22/files#diff-f857355bf8f8ba6545f5a75723b994d6cbbde67452c690c391add18bd42dfe9d

and see if this looks correct to you? (Don't worry about the other changes in that PR; I'll get an internal team review for those.) Note that I haven't tried running this yet so there may be some obvious bugs. Also note that this is a queue worker so it has a bit of an odd structure, using a bunch of global variables, since that's how startup initialization is cached given how the worker code is run.

Could you provide me with a stencil query and data ID (Butler UUID) that you would expect to retrieve a meaningful cutout with the current IDF data set? I will need that to do integration testing.

Also, forgive a possibly stupid question about how the stack works: should I expect `lsst.image_cutout_backend` to magically appear as a stack package in weekly builds, or do I have to do something special to ensure that it's installed in a stack container?

**Comment by TallJimbo** [comment]
> Could you provide me with a stencil query and data ID (Butler UUID) that you would expect to retrieve a meaningful cutout with the current IDF data set? I will need that to do integration testing.

I just took a quick stab at this using the DP0.1 RSP, and realized that we've both assumed that we'd be using UUIDs for dataset IDs, but that was actually a change we made after DP0.1 was frozen, so we can't use that repo.  I'm sure there's something we can use from the DP0.2 processing that's already completed, but I'll need to go look up the butler repo configuration for that; hopefully I can get to that tomorrow.

> Also, forgive a possibly stupid question about how the stack works: should I expect `lsst.image_cutout_backend` to magically appear as a stack package in weekly builds, or do I have to do something special to ensure that it's installed in a stack container?

It won't appear magically, I'm afraid.  After the usual stack environment setup (not sure how that works in containers, actually, but I assume it involves sourcing `loadLSST.sh` or something), you'll need to do

```
git clone https://github.com/lsst-dm/image_cutout_backend.git
cd image_cutout_backend
setup -r .
scons install declare -t current
```
once, and then in any later use of the container you should be able to do
```
setup image_cutout_backend
```
If that doesn't work (or if it's a pain to try out), we may want to get input from somebody who knows about both the container stack installs and EUPS, as I think we've each only got half of the picture.

**Review by rra — APPROVED** [review]
I am not qualified to review the implementation, but in terms of the interface this all looks good to me.

**Comment by TallJimbo** [comment]
Here's a circular stencil (degrees):
```
ra: 53.13925
dec: -34.0215
radius: 0.0105
```
and a dataset UUID from the DP0.2 processing that contains it:
```
8a953c0321bd4878bfa694dbf628ea81
```
I put this together on a `data-int` notebook, pointing at this butler repository configuration:
```
s3://butler-us-central1-panda-dev/dc2/butler-external.yaml
```
I'm not sure if that's an appropriate one for your integration test, but if not I hope there's another view that might be.

If not, or we need to debug something, this is a `calexp` in collection
```
2.2i/runs/DP0.2/v23_0_0_rc5/PREOPS-905
```
with data ID
```
{'visit': 192350, 'detector': 175, 'band': 'i'}
```
(it's the same one as in `tutorial-notebooks/03b_Image_Display_with_Firefly.ipynb`).

We should be able to use those to find a UUID compatible with this stencil in any data repository with similar processing, or do that processing ourselves if we want to make a small test data repository and don't want to export from another one.

**Comment by rra** [comment]
Thanks for the test case! The code currently on this branch fails with that test case with the following error, which I think is not at my end:
```
[2022-02-05 00:22:45,757] [PID 610] [Thread-4] [dramatiq.worker.WorkerThread] [ERROR] Failed to process message cutout('2', ['8a953c0321bd4878bfa694dbf628ea81'], [{'type': 'circle', 'center': {'ra': 53.13925, 'dec': -34.0215}, 'radius': 0.0105}]) with unhandled exception.
Traceback (most recent call last):
  File "/tmp/vo-cutouts/src/vocutouts/./workers.py", line 153, in cutout
    result = backend.process_uuid(sky_stencils[0], UUID(dataset_ids[0]))
  File "/tmp/image_cutout_backend/python/lsst/image_cutout_backend/_image_cutout_backend.py", line 225, in process_uuid
    extract_result = self.extract_uuid(stencil, uuid, component=component)
  File "/tmp/image_cutout_backend/python/lsst/image_cutout_backend/_image_cutout_backend.py", line 343, in extract_uuid
    return self.extract_ref(stencil, ref)
  File "/tmp/image_cutout_backend/python/lsst/image_cutout_backend/_image_cutout_backend.py", line 292, in extract_ref
    wcs, bbox = self.projection_finder(ref, self.butler)
  File "/tmp/image_cutout_backend/python/lsst/image_cutout_backend/projection_finders.py", line 101, in __call__
    raise LookupError(f"No way to obtain WCS and bounding box information for ref {ref}.")
LookupError: No way to obtain WCS and bounding box information for ref calexp@{instrument: 'LSSTCam-imSim', detector: 175, visit: 192350}, sc=ExposureF] (id=8a953c03-21bd-4878-bfa6-94dbf628ea81).
```

**Comment by TallJimbo** [comment]
Just pushed a commit to the branch that should fix it.

**Comment by rra** [comment]
Yup, that fixed that! New error:
```
[2022-02-05 02:21:37,501] [PID 628] [Thread-5] [dramatiq.worker.WorkerThread] [ERROR] Failed to process message cutout('3', ['8a953c0321bd4878bfa694dbf628ea81'], [{'type': 'circle', 'center': {'ra': 53.13925, 'dec': -34.0215}, 'radius': 0.0105}]) with unhandled exception.
Traceback (most recent call last):
  File "/tmp/vo-cutouts/src/vocutouts/./workers.py", line 153, in cutout
    result = backend.process_uuid(sky_stencils[0], UUID(dataset_ids[0]))
  File "/tmp/image_cutout_backend/python/lsst/image_cutout_backend/_image_cutout_backend.py", line 225, in process_uuid
    extract_result = self.extract_uuid(stencil, uuid, component=component)
  File "/tmp/image_cutout_backend/python/lsst/image_cutout_backend/_image_cutout_backend.py", line 343, in extract_uuid
    return self.extract_ref(stencil, ref)
  File "/tmp/image_cutout_backend/python/lsst/image_cutout_backend/_image_cutout_backend.py", line 294, in extract_ref
    pixel_stencil = stencil.to_pixels(wcs, bbox)
  File "/tmp/image_cutout_backend/python/lsst/image_cutout_backend/stencils.py", line 275, in to_pixels
    return self.to_polygon(n_vertices).to_pixels(wcs, bbox)
  File "/tmp/image_cutout_backend/python/lsst/image_cutout_backend/stencils.py", line 268, in to_polygon
    return SkyPolygon(self._center.offset(b * factor, self._radius) for b in range(n_vertices))
TypeError: 'numpy.float64' object cannot be interpreted as an integer
```

**Comment by TallJimbo** [comment]
Another fix pushed.  I guess it's good that your test happens to be trying all of the code paths that my one-off test didn't, but it does make me wish it was easier to write those unit tests.  Sorry it's been a pain.

**Comment by rra** [comment]
No problem! Okay, that got me up to permission errors with the S3 code because writing the object also needs read permissions. I thought that might be the case and should have dealt with that originally. I'll fix that on Monday since that requires a code review from someone else.

**Comment by rra** [comment]
After a few more fixes on my end for the storage and retrieval parts, I have gotten this working end-to-end. Working test URL:

https://data-int.lsst.cloud/api/cutout/sync?id=8a953c0321bd4878bfa694dbf628ea81&circle=53.13925%20-34.0215%200.0105

The only remaining problem that I saw with the backend is that you're not deleting the temporary file on disk after uploading it to S3. Could you please add that? Otherwise, the worker pods are going to run out of local memory fairly quickly. (These run in a container, so there's no such thing as `/tmp` cleaning or the like.)

**Comment by timj** [comment]
It's meant to delete the temporary file. That's what this code is meant to do:

```python
with ResourcePath.temporary_uri(prefix=self.temporary_root, suffix=".fits") as tmp_uri:
    extract_result.write_fits(tmp_uri.ospath)
    remote_uri.transfer_from(tmp_uri, transfer="copy")
```

**Comment by rra** [comment]
It looks like after each backend execution I get a new directory in `/tmp` named `butler-` followed by some random letters and containing two files:
```
8a953c03-21bd-4878-bfa6-94dbf628ea81.fits  exempt
```

**Comment by timj** [comment]
> It looks like after each backend execution I get a new directory in /tmp

That's the butler read cache. When you do a butler.get we stuff the file in local tmp so that if you butler.get it immediately (or get a component) it will read the local file. This makes a huge speed up for some operations in notebooks or for pipeline execution if you chain some tasks together since the next task will read the local file.

That's all controlled in the datastore configuration: https://github.com/lsst/daf_butler/blob/main/python/lsst/daf/butler/configs/datastores/fileDatastore.yaml#L14-L39

You can have a local override of that by setting a `$DAF_BUTLER_CONFIG_PATH` environment variable and writing a `datastores/fileDatastore.yaml` file in that directory with `cached.expiry.mode: null` in it (in yaml syntax).

**Comment by rra** [comment]
I wouldn't mind a cache but it seems to make a new one every time? Although maybe that's because there are multiple workers. Does it ever clean up stale entries to keep its disk usage in check, or would it eventually just run out of space?

**Comment by rra** [comment]
Ack, sorry, should have followed the link. Okay, I think I actually want that cache, so I think we're good!

**Comment by timj** [comment]
For all I know you want to tweak it to use a fixed temp location and cache size rather than dataset count.

**Comment by TallJimbo** [comment]
@timj , thanks for explaining that, as it would have taken me a while to figure out what was actually going on.

@rra , if you're happy with this, I'm happy to merge, but I see that you haven't hit Reviewed in Jira, and I wasn't sure if that was intentional or not given the ongoing testing.

**Comment by rra** [comment]
I was holding off until I got a clean end-to-end test, which I didn't get until late yesterday, but everything looks good now. I have marked the issue as reviewed.

**Pull request merged**
